src/mock/quick_p2p/tests.rs

```diff
@@ -0,0 +1,542 @@
+    let mut b = Agent::node(&network);
+
+    establish_connection(&network, &mut a, &mut b);
+
+    b.disconnect_from(a.addr());
+    network.poll();
+
+    a.expect_connection_failure(&b.addr());
+    b.expect_none();
+}
+
+#[test]
+fn disconnect_incoming_connection() {
+    let network = Network::new(None);
+
+    let mut a = Agent::node(&network);
+    let mut b = Agent::node(&network);
+
+    establish_connection(&network, &mut a, &mut b);
+
+    a.disconnect_from(b.addr());
+    network.poll();
+
+    a.expect_none();
+    b.expect_connection_failure(&a.addr());
+}
+
+#[test]
+fn send_to_connected_node() {
+    let network = Network::new(None);
+
+    let mut a = Agent::node(&network);
+    let mut b = Agent::node(&network);
+
+    establish_connection(&network, &mut a, &mut b);
+
+    let msg = Bytes::from_static(b"message from A");
+    a.send(b.addr(), msg.clone());
+    network.poll();
+
+    b.expect_new_message(&a.addr(), &msg);
+}
+
+#[test]
+fn send_to_disconnecting_node() {
+    let network = Network::new(None);
+    let mut a = Agent::node(&network);
+    let mut b = Agent::node(&network);
+
+    establish_connection(&network, &mut a, &mut b);
+
+    let msg = Bytes::from_static(b"message from A");
+    a.send(b.addr(), msg.clone());
+    b.disconnect_from(a.addr());
+    network.poll();
+
+    a.expect_connection_failure(&b.addr());
+    a.expect_unsent_message(&b.addr(), &msg);
+    b.expect_none();
+}
+
+#[test]
+fn send_to_nonexisting_node() {
+    let network = Network::new(None);
+
+    let mut a = Agent::node(&network);
+    let b_addr = network.gen_addr();
+
+    let msg = Bytes::from_static(b"message from A");
+    a.send(b_addr, msg.clone());
+    network.poll();
+
+    // Note: the real quick-p2p will only emit `UnsentUserMessage` when a connection to the peer
+    // was previously successfully established. That is not the case here, so we expect nothing.
+    a.expect_none();
+}
+
+#[test]
+fn send_without_connecting_first() {
+    let network = Network::new(None);
+    let mut a = Agent::node(&network);
+    let b = Agent::node(&network);
+
+    let msg = Bytes::from_static(b"message from A");
+    a.send(b.addr(), msg.clone());
+
+    network.poll();
+
+    a.expect_connected_to_node(&b.addr());
+    b.expect_connected_to_node(&a.addr());
+    b.expect_new_message(&a.addr(), &msg);
+}
+
+#[test]
+fn send_multiple_messages_without_connecting_first() {
+    let network = Network::new(None);
+    let mut a = Agent::node(&network);
+    let b = Agent::node(&network);
+
+    let msg0 = Bytes::from_static(b"message 0 from A");
+    a.send(b.addr(), msg0.clone());
+
+    let msg1 = Bytes::from_static(b"message 1 from A");
+    a.send(b.addr(), msg1.clone());
+
+    let msg2 = Bytes::from_static(b"message 2 from A");
+    a.send(b.addr(), msg2.clone());
+
+    network.poll();
+
+    a.expect_connected_to_node(&b.addr());
+    b.expect_connected_to_node(&a.addr());
+
+    b.expect_new_message(&a.addr(), &msg0);
```

Comment:
Is that always true? No re-ordering?

Reply:
Within the same node, `ConnectedTo` is guaranteed to be raised before `NewMessage`. There is no guarantee on the order different nodes will receive the messages, but that doesn't matter anyway.

Reply:
I mean is the mock doing no re-ordering while the real code may do so?
Is that correct?
Is that not a problem?

Reply:
There is randomness in the order messages are processed (see `Network::poll`) which is supposed to simulate the uncertain order in the real world. So for example, when we have something like this:
```rust
a.send(c, "msg0");
b.send(c, "msg1");
```
then there is no guarantee which message will be delivered first. This non-guarantee is followed in this mock code too. 

However, in the case of this test, the `ConnectedTo`  is guaranteed to be raised before any `NewMessage` from the same peer. This is the case in the real code too.

Reply:
That is a different example than the test case:

```
a.send(b, "msg0");
a.send(b, "msg1");
```
Can these messages be seen in a different order by b?
in the real world?
in the tests?

Reply:
As, I see what you mean now. So yes, in that case they can be reordered in the real world, but currently not in mock. Do you think it can be a problem? Is it worth implementing the reordering in the mock too?

Reply:
I think it will, so our test can detect these issues with mock. However, I'm happy for it to be a follow up. Maybe leave a TODO. (Could be something someone else does).